community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NumberKey.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema;
+
+import org.neo4j.index.internal.gbptree.GBPTree;
+
+/**
+ * Includes value and entity id (to be able to handle non-unique values).
+ * A value can be any {@link Number} and is represented as a {@code long} to store the raw bits and a type
+ * to say if it's a long, double or float.
+ *
+ * Distinction between double and float exists because coersions between each other and long may differ.
+ * TODO this should be figured out and potentially reduced to long, double types only.
+ */
+class NumberKey
+{
+    static final int SIZE =
+            Byte.BYTES + /* type of value */
+            Long.BYTES + /* raw value bits */
+
+            // TODO this could use 6 bytes instead and have the highest 2 bits stored in the type byte
+            Long.BYTES;  /* entityId */
+
+    static final byte TYPE_LONG = 0;
+    static final byte TYPE_FLOAT = 1;
+    static final byte TYPE_DOUBLE = 2;
+
+    byte type;
+    long rawValueBits;
+    long entityId;
+
+    /**
+     * Marks that comparisons with this key requires also comparing entityId, this allows functionality
+     * of inclusive/exclusive bounds of range queries.
+     * This is because {@link GBPTree} only support from inclusive and to exclusive.
+     * <p>
+     * Note that {@code entityIdIsSpecialTieBreaker} is only an in memory state.
+     */
+    boolean entityIdIsSpecialTieBreaker;
+
+    void from( long entityId, Object[] values )
+    {
+        assertValidSingleNumberPropertyValue( values );
+        extractValue( (Number) values[0] );
+        this.entityId = entityId;
+        entityIdIsSpecialTieBreaker = false;
+    }
+
+    String propertiesAsString()
+    {
+        return String.valueOf( toNumberValue() );
+    }
+
+    void initAsLowest()
+    {
+        rawValueBits = Double.doubleToLongBits( Double.NEGATIVE_INFINITY );
```

Comment:
Lucene uses `Double.compare` so that behaviour should be similar at least.
As for highest/lowest, this is only a way for us to create a range that is unbound (left, right or both) and because we use `Double.compare` for comparison it should not be possible to store anything outside of this limit.
Do you still we need to define a common place for this?